lib/dockerregistry/inventory.go

```diff
@@ -404,6 +395,92 @@ func (sc *SyncContext) getPromotionCandidates(
 	return toPromote
 }
 
+// checkOverlappingEdges checks to ensure that all the edges taken together as a
+// whole are consistent. It checks that there are no duplicate promotions
+// desired to the same destination vertex (same destination PQIN). If the
+// digests are the same for those edges, ignore because by definition the
+// digests are cryptographically guaranteed to be the same thing (it doesn't
+// matter if 2 different parties want the same exact image to become promoted to
+// the same destination --- in many ways this is actually a good thing because
+// it's a form of redundancy). However, return an error if the digests are
+// different, because most likely this is at best just human error and at worst
+// a malicious attack (someone trying to push an image to an endpoint they
+// shouldn't own, or possibly an attempt to rename a tag for an existin image to
+// be something else).
+//
+// nolint[gocyclo]
+func checkOverlappingEdges(
+	edges map[PromotionEdge]interface{}) (map[PromotionEdge]interface{}, error) {
+
+	// Build up a "promotionIntent". This will be checked below.
+	promotionIntent := make(map[string]map[Digest][]PromotionEdge)
+	for edge := range edges {
+		dstPQIN := ToPQIN(edge.DstRegistry.Name,
+			edge.DstImageTag.ImageName,
+			edge.DstImageTag.Tag)
+
+		digestToEdges, ok := promotionIntent[dstPQIN]
+		if ok {
+			// Store the edge in the appropriate bucket.
+			edgeList, ok := digestToEdges[edge.Digest]
+			if ok {
+				// There is already a bucket for edges that have the same digest
+				// as this edge, so just append to that list.
+				edgeList = append(edgeList, edge)
+				digestToEdges[edge.Digest] = edgeList
+			} else {
+				digestToEdges[edge.Digest] = []PromotionEdge{edge}
+			}
+			promotionIntent[dstPQIN] = digestToEdges
+		} else {
+			// Make this edge lay claim to this destination vertex.
+			edgeList := make([]PromotionEdge, 0)
+			edgeList = append(edgeList, edge)
+			digestToEdges := make(map[Digest][]PromotionEdge)
+			digestToEdges[edge.Digest] = edgeList
+			promotionIntent[dstPQIN] = digestToEdges
+		}
+	}
+
+	// Review the promotionIntent to ensure that there are no issues.
+	overlapError := false
+	checked := make(map[PromotionEdge]interface{})
+	for pqin, digestToEdges := range promotionIntent {
+		if len(digestToEdges) < 2 {
+			for _, edgeList := range digestToEdges {
+				switch len(edgeList) {
+				case 0:
+					klog.Errorf("no edges for %v\n", pqin)
```

Comment:
Fatalf?  Something's gone seriously wrong here, right?

(Also you don't need the trailing newline)

Reply:
Oops, I should set `overlapError = true` here. Yes something's gone seriously wrong, but I want to collect as much error as possible so I would like to avoid Fatalf here.

Hmm I don't get why I added the trailing newlines (looks like I forgot this is not C again). Ack.

Reply:
Hmm actually it should be a different type of error, because it's not like an overlap was actually detected. Will change.